### ISSUE TITLE
TensorRT batch inference

### ISSUE THREAD

**Issue by erdongchendou** [ultralytics/ultralytics#7125]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

I export the yolov8 detection model with the following code.
``` python 
import os
from ultralytics import YOLO

model = YOLO("./yolov8l.pt")
model.export(format="engine", imgsz=640, device="0", dynamic=True, batch=6)
``` 

Do inference with the following code.

``` python
import torch

model = YOLO("./yolov8l.engine") 
source = torch.rand(6, 3, 640, 640, dtype=torch.float32)
results = model.predict(source, save=False, batch=6, device="0")
```

And I get the following log, it seems like trt yolov8 does not do batch inference but do inference on one image a time.

0: 640x640 (no detections), 1: 640x640 (no detections), 2: 640x640 (no detections), 
3: 640x640 (no detections), 4: 640x640 (no detections), 5: 640x640 (no detections), 
194.0ms
Speed: 0.5ms preprocess, 32.3ms inference, 6.6ms postprocess per image at shape (1, 3, 640, 640)
Do I do something wrong?

### Environment

```
Ultralytics YOLOv8.0.219 🚀 Python-3.9.18 torch-1.8.2+cu111 CUDA:0 (Tesla T4, 15110MiB)
Setup complete ✅ (64 CPUs, 125.6 GB RAM, 1490.4/7096.5 GB disk)

OS                  Linux-5.4.0-146-generic-x86_64-with-glibc2.31
Environment         Docker
Python              3.9.18
Install             pip
RAM                 125.56 GB
CPU                 Intel Xeon Silver 4314 2.40GHz
CUDA                11.1

matplotlib          ✅ 3.8.1>=3.3.0
numpy               ✅ 1.23.5>=1.22.2
opencv-python       ✅ 4.8.1.78>=4.6.0
pillow              ✅ 10.1.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.3>=1.4.1
torch               ✅ 1.8.2+cu111>=1.8.0
torchvision         ✅ 0.9.2+cu111>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 2.1.2>=1.1.4
seaborn             ✅ 0.13.0>=0.11.0
psutil              ✅ 5.9.6
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
```

### Additional

The following is the export log.
> Ultralytics YOLOv8.0.219 🚀 Python-3.9.18 torch-1.8.2+cu111 CUDA:0 (Tesla T4, 15110MiB)
YOLOv8l summary (fused): 268 layers, 43668288 parameters, 0 gradients, 165.2 GFLOPs
> 
> PyTorch: starting from '/hmty/ultralytics/yolov8l.pt' with input shape (6, 3, 640, 640) BCHW and output shape(s) (6, 84, 8400) (83.7 MB)
> 
> ONNX: starting export with onnx 1.15.0 opset 12...
> /usr/local/lib/python3.9/site-packages/torch/tensor.py:587: RuntimeWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
>   warnings.warn('Iterating over a tensor might cause the trace to be incorrect. '
> ONNX: simplifying with onnxsim 0.4.35...
> ONNX: export success ✅ 9.9s, saved as '/hmty/ultralytics/yolov8l.onnx' (166.6 MB)
> 
> TensorRT: starting export with TensorRT 8.4.3.1...
> [12/22/2023-01:24:56] [TRT] [I] [MemUsageChange] Init CUDA: CPU +302, GPU +0, now: CPU 3645, GPU 6698 (MiB)
> [12/22/2023-01:24:57] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +206, GPU +72, now: CPU 3871, GPU 6770 (MiB)
> [12/22/2023-01:24:57] [TRT] [I] ----------------------------------------------------------------
> [12/22/2023-01:24:57] [TRT] [I] Input filename:   /hmty/ultralytics/yolov8l.onnx
> [12/22/2023-01:24:57] [TRT] [I] ONNX IR version:  0.0.6
> [12/22/2023-01:24:57] [TRT] [I] Opset version:    12
> [12/22/2023-01:24:57] [TRT] [I] Producer name:    pytorch
> [12/22/2023-01:24:57] [TRT] [I] Producer version: 1.8
> [12/22/2023-01:24:57] [TRT] [I] Domain:           
> [12/22/2023-01:24:57] [TRT] [I] Model version:    0
> [12/22/2023-01:24:57] [TRT] [I] Doc string:       
> [12/22/2023-01:24:57] [TRT] [I] ----------------------------------------------------------------
> [12/22/2023-01:24:57] [TRT] [W] onnx2trt_utils.cpp:369: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
> TensorRT: input "images" with shape(-1, 3, -1, -1) DataType.FLOAT
> TensorRT: output "output0" with shape(-1, 84, -1) DataType.FLOAT
> TensorRT: building FP32 engine as /hmty/ultralytics/yolov8l.engine
> [12/22/2023-01:24:59] [TRT] [I] [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +512, GPU +244, now: CPU 4568, GPU 6362 (MiB)
> [12/22/2023-01:25:00] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +118, GPU +76, now: CPU 4686, GPU 6438 (MiB)
> [12/22/2023-01:25:00] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:25:00] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
> [12/22/2023-01:25:43] [TRT] [I] Some tactics do not have sufficient workspace memory to run. Increasing workspace size will enable more tactics, please check verbose output for requested sizes.
> 
> 
> [12/22/2023-01:27:19] [TRT] [I] Detected 1 inputs and 3 output network tensors.
> [12/22/2023-01:27:20] [TRT] [I] Total Host Persistent Memory: 147328
> [12/22/2023-01:27:20] [TRT] [I] Total Device Persistent Memory: 1532416
> [12/22/2023-01:27:20] [TRT] [I] Total Scratch Memory: 27070976
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 0 MiB, GPU 0 MiB
> [12/22/2023-01:27:20] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 137.348ms to assign 13 blocks to 273 nodes requiring 579994120 bytes.
> [12/22/2023-01:27:20] [TRT] [I] Total Activation Memory: 579994120
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +0, GPU +34, now: CPU 5205, GPU 6878 (MiB)
> [12/22/2023-01:27:20] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in building engine: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> [12/22/2023-01:27:20] [TRT] [W] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
> [12/22/2023-01:27:20] [TRT] [W] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
> TensorRT: export success ✅ 155.4s, saved as '/hmty/ultralytics/yolov8l.engine' (190.7 MB)
> 
> Export complete (161.7s)
> Results saved to /hmty/ultralytics
> Predict:         yolo predict task=detect model=/hmty/ultralytics/yolov8l.engine imgsz=640  
> Validate:        yolo val task=detect model=/hmty/ultralytics/yolov8l.engine imgsz=640 data=coco.yaml  
> Visualize:       https://netron.app

The following is the inference log.

> WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify', or 'pose'.
> Loading yolov8l.engine for TensorRT inference...
> [12/22/2023-01:31:52] [TRT] [I] [MemUsageChange] Init CUDA: CPU +310, GPU +0, now: CPU 407, GPU 4548 (MiB)
> [12/22/2023-01:31:52] [TRT] [I] Loaded engine size: 190 MiB
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +629, GPU +292, now: CPU 1262, GPU 5034 (MiB)
> [12/22/2023-01:31:53] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +0, GPU +32, now: CPU 1071, GPU 5034 (MiB)
> [12/22/2023-01:31:53] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> 
> 0: 640x640 (no detections), 1: 640x640 (no detections), 2: 640x640 (no detections), 3: 640x640 (no detections), 4: 640x640 (no detections), 5: 640x640 (no detections), 238.2ms
> Speed: 0.6ms preprocess, 39.7ms inference, 5.2ms postprocess per image at shape (1, 3, 640, 640)

The code I use is the example code from here: https://docs.ultralytics.com/modes/export/#arguments 
From my experience and knowledge, I think the export process is successful and I set batch size at inference time. So I can not find where the problem is. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@erdongchendou hello there! 👋 It looks like your inquiry is about batch inference with a TensorRT engine exported from YOLOv8. From the logs you've provided, you're right to conclude that the model export appears successful, and you're also setting the batch size for inference correctly.

The message you're seeing in the logs, regarding inferring one image at a time, may not necessarily indicate a problem. It could simply be the way logging is structured – reporting each item in the batch sequentially. However, if you're experiencing significantly longer inference times than expected for batch processing, this could suggest that the model isn't utilizing batch inference effectively.

The key to ensure that TensorRT uses batch inference is during the export process, which you've done by setting the `batch` parameter in the model export function. If everything was set correctly there, the engine should be capable of performing batch inferences.

If you're still concerned that batch inference isn't functioning as expected, one step is to review the logs during export to verify that dynamic shapes were correctly set for batching. You'd want to see input shapes like `(-1, 3, -1, -1)`, which indicate that the model supports dynamic batching. Additionally, check your device utilization during inference to confirm whether multiple inputs are being processed in parallel.

For further technical support, consider sharing this issue and any additional relevant details on our GitHub Issues page, and our team will attempt to provide more insight. You can also refer to the comprehensive export documentation provided at https://docs.ultralytics.com for more details about model exporting.

Keep exploring, and thanks for using YOLOv8! 🚀

**Comment by erdongchendou** [comment]
From the log I already provide, the input size is (-1, 3, -1, -1).
``` bash
[12/22/2023-01:24:57] [TRT] [W] onnx2trt_utils.cpp:369: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
TensorRT: input "images" with shape(-1, 3, -1, -1) DataType.FLOAT
TensorRT: output "output0" with shape(-1, 84, -1) DataType.FLOAT
TensorRT: building FP32 engine as /hmty/ultralytics/yolov8l.engine
```
I ran the inference code for 100 times, and the total time used is 30 seconds. So the inference time for one batch is 300ms， which means that the TensorRT inference is processing one image a time.

The following is the code I use.
``` python 

import time
import torch
import datetime

model = YOLO("yolov8l.engine")  # load an official model
source = torch.rand(6, 3, 640, 640, dtype=torch.float32)
t0 = time.time()
print("start time is ", datetime.datetime.now())
for i in range(100):
    results = model.predict(source, save=False, batch=6, device="1", verbose=False)
t1 = time.time()
print("end time is ", datetime.datetime.now())
print(t1 - t0)

```

The following is the log from running above code.

``` bash
root@ubuntu:/hmty/ultralytics# python scripts/detection_inference.py 
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify', or 'pose'.
start time is  2023-12-25 02:52:37.824872
Loading yolov8l.engine for TensorRT inference...
[12/25/2023-02:52:38] [TRT] [I] [MemUsageChange] Init CUDA: CPU +310, GPU +0, now: CPU 407, GPU 6396 (MiB)
[12/25/2023-02:52:38] [TRT] [I] Loaded engine size: 198 MiB
[12/25/2023-02:52:39] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +628, GPU +292, now: CPU 1270, GPU 6890 (MiB)
[12/25/2023-02:52:39] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
[12/25/2023-02:52:39] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
[12/25/2023-02:52:39] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +0, GPU +32, now: CPU 1072, GPU 6890 (MiB)
[12/25/2023-02:52:39] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
[12/25/2023-02:52:39] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
end time is  2023-12-25 02:53:08.769049
30.944129467010498
```
